platform-hub-api/app/models/kubernetes_token.rb

```diff
@@ -1,18 +1,141 @@
+  scope :by_tokenable, ->(tokenable) { 
+    where("tokenable_type=? AND tokenable_id=?", tokenable.class.name, tokenable.to_param) 
+  }
+  scope :by_cluster, ->(c) { where(:cluster_id => c.id) }
+  scope :by_name, ->(n) { where(:name => n.downcase) }
+
+  enum kind: {
+    user: 'user',
+    robot: 'robot',
+  }
+
+  validates :uid, presence: true, length: { is: UID_LENGTH }, uniqueness: true
+  validates :kind, presence: true
+  validates :name,
+    presence: true,
+    format: {
+      with: NAME_REGEX,
+      message: "must start with letter and can only contain letters, numbers, underscores, dashes, dots and @"
+    }
+  validates :description, presence: true, if: :robot?
+
+  validate :tokenable_type_set_properly
+  validate :token_must_not_be_blank
+  validate :token_must_be_of_expected_length
+  validate :one_user_token_per_cluster
+  validate :robot_name_unique_for_given_cluster
+
+  def token=(val)
+    self['token'] = ENCRYPTOR.encrypt(val)
+  end
+
+  def groups=(groups)
+    self['groups'] = 
+      if groups.is_a? String
+        groups.split(',').map(&:strip).reject(&:blank?).uniq
+      elsif groups.is_a? Array
+        groups.uniq
+      end
+  end
+
+  def decrypted_token
+    ENCRYPTOR.decrypt(token)
+  end
+
+  def user
+    if robot?
+      tokenable
+    elsif user? # via identity
+      tokenable.user
+    end
+  end
+
+  def privileged?
+    expire_privileged_at.present?
+  end
+
+  def escalate(privileged_group, expire_in_secs = 600)
+    update_attributes(
+      expire_privileged_at: [ expire_in_secs, PRIVILEGED_GROUP_MAX_EXPIRATION_SECONDS ].min.seconds.from_now,
+      groups: groups << privileged_group
+    )
+  end
+
+  def deescalate
+    update_attributes(
+      expire_privileged_at: nil, 
+      groups: groups - KubernetesGroup.privileged_names
+    )
+  end
+
+  protected
+
+  def tokenable_type_set_properly
+    if robot?
+      errors.add(:tokenable_type, "must be `User` for robot token") if tokenable_type != 'User'
+    elsif user?
+      errors.add(:tokenable_type, "must be `Idendity` for user token") if tokenable_type != 'Identity'
+    end
+  end
+
+  def token_must_not_be_blank
+    if decrypted_token.blank?
+      errors.add(:token, "can't be blank")
+    end
+  end
+
+  def token_must_be_of_expected_length
+    if decrypted_token.present? && decrypted_token.length != TOKEN_LENGTH
+      errors.add(:token, "is the wrong length (should be #{TOKEN_LENGTH} characters)")
+    end
+  end
+
+  def one_user_token_per_cluster
+    # For user token we only allow one per cluster
+    return unless user?
+    if new_record? && user_token_for_cluster_exists?(cluster)
+      errors.add(:user, "can have only one user token per cluster")
+    end
+  end
+
+  def robot_name_unique_for_given_cluster
+    # For robot token we allow multiple per cluster unique by name
+    return unless robot?
+    if new_record? && robot_token_for_cluster_and_name_exists?(cluster, name)
```

Comment:
If we do make the `name` editable, then this will need updating (so it also runs on updates).

Reply:
Why? Since `name` and `cluster` don't change this only need to be triggered on creation.

Reply:
Yeah, my comment was connected to the suggestion before of making `name` modifiable. But if we're keeping it immutable then no need to change this.